### PR TITLE
PReLU cuda implementation

### PR DESCRIPTION
**Pull request by adampolyak**
Added cuda support for PReLU jointly with @szagoruyko.
Requires nn pull request torch/nn#258 to work.

### CODE REVIEW

**Comment by szagoruyko** [comment]
it might be possible to remove one buffer if there was THCudaTensor_reduceDim which allows an additional functor to modify output.
another thing is backward, we made it reuse gradInput and them recompute it to restore the state, because otherwise it would require an additional buffer size of the input, or lot's of big kernels which do reduceDim but take two tensors on input.

**Comment by soumith** [comment]
you need to rebase the prelu commits back to top of master. if you notice, they are mixed, and the current PR shows commits that are already merged into trunk as well.

**Comment by adampolyak** [comment]
fixed

File: PReLU.cu [hunk 1]
```diff
@@ -0,0 +1,263 @@
+#include "THCApply.cuh"
+#include "THCReduce.cuh"
+#include "utils.h"
+
+// This is bad, because the following is defined in SpatialConvolution.cu ...
```
Comment (soumith) [inline PReLU.cu]: can you #undef at the end of SpatialConvolution.cu , and uncomment these

Comment (szagoruyko) [inline PReLU.cu]: btw would it be better to separate the compilation of each file in cunn? take some time now

Comment (dominikgrewe) [inline PReLU.cu]: I agree we should compile each file separately. We do it in cutorch now, which enables us to compile files in parallel.

Comment (nicholas-leonard) [inline PReLU.cu]: How do you accomplish that? Curious.

Comment (soumith) [inline PReLU.cu]: @nicholas-leonard https://github.com/torch/demos/blob/master/train-on-housenumbers/train-on-housenumbers.lua#L243
$(MAKE) -j$(getconf _NPROCESSORS_ONLN) install

**Comment by adampolyak** [comment]
Added undef where required. Thing is, we are still using CUDA_NUM_THREADS and GET_BLOCKS which are defined in SpatialConvolutionMM.cu. By the way, VolumetricConvolution is also using those.

**Comment by soumith** [comment]
ok, i'll have to do a cleanup pass on the macro defs, and splitting the compilation.

**Comment by soumith** [comment]
if you squash your commits now, its ready to merge.

**Comment by adampolyak** [comment]
done

**Comment by soumith** [comment]
Thanks Adam!

**Pull request merged**